Add tag app release action

sarayourfriend opened the pull request:
## Fixes

<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->

Fixes #976 by @sarayourfriend 

## Description

<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
This PR adds a new `Release app` workflow. This workflow requires the choice of an app (api, ingestion server, or frontend) and an existing image tag for that app. It then does the following:

1. Generate a new release tag based on the UTC timestamp when the workflow is dispatched
2. Create a new docker image tag based on the existing image(s) with the timestamp (`rel-{timestamp}`). The docker image tag does not include the app name as it'd be redundant with the rest of the image name
3. Creates and publishes a release with the changes since the previous release. Until each application has it's own release generated by this workflow, they will all the last "General" release as the "previous release". Currently that is v4.0.1. Once each of them have their own releases from this workflow, however, from then on they will derive the changes between the previous release for that app only up until the commit corresponding to the docker image SHA.
   - This step also creates the git tag. We do not need to manually do it ourselves as creating a GitHub release also creates the git tag
4. Pulls the text from the release (the changelog) and generates a Markdown file to add to the documentation site that includes all the changelogs (starting with when we begin using this process).
5. If run for the API or frontend, then it also dispatches the corresponding production deployment workflow

## Testing Instructions

<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
This was tested extensively on this branch. A summary of that can be found here: https://github.com/WordPress/openverse/pull/987#issuecomment-1482270090

## Checklist

<!-- Replace  the [ ] with [x] to check the boxes. -->

- [x] My pull request has a descriptive title (not a vague title like
      `Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or
      a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [N/A] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [N/A] I tried running the project locally and verified that there are no visible
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

AetherUnbound commented on .github/release_drafter_api.yml:
```diff
@@ -0,0 +1,30 @@
+  - title: Bug Fixes
+    label: "🛠 goal: fix"
+change-template: "- $TITLE (#$NUMBER) @$AUTHOR"
+include-labels:
+  - "🧱 stack: api"
```
Being able to make this distinction is SO COOL, WOW! 🤯

AetherUnbound commented on .github/workflows/tag-app-release.yml:
```diff
@@ -0,0 +1,123 @@
+on:
+  workflow_dispatch:
+    inputs:
+      app:
+        type: string
+        required: true
+        description: Application to tag for. Either `ingestion_server`, `api`, or `frontend`.
```
Could this potentially be of type `choice`? [Example](https://github.blog/changelog/2021-11-10-github-actions-input-types-for-manual-workflows/). That way we wouldn't have to let this be a free-form field and perform the validation step below explicitly.

sarayourfriend replied:
Yes, definitely. I totally overlooked that.

AetherUnbound commented on documentation/changelogs/api/2023.03.22.04.56.29.md:
```diff
@@ -0,0 +1,33 @@
+- Update to new link for pipenv install instructions (#962) @kk311y
+- Treat any non 200 status as failure for thingiverse (#940) @sarayourfriend
+- Skip build and publish job if nothing to do (#977) @dhruvkb
+- Make `searchTerm` for VAudioTrack and VImageCell optional (#955) @obulat
```
Should this be part of the API changelog? It doesn't appear to edit any API files nor does it have the API stack label 🤔

sarayourfriend replied:
As noted in the commit that adds these files, they're just here for testing: https://github.com/WordPress/openverse/pull/987/commits/1a757d19bee7b4f57068a48e105132affe81355f

I copied the file from the existing draft release, it has nothing to do with the release-drafter workflow configurations introduced in this PR.

AetherUnbound commented on documentation/changelogs/api/2023.03.22.04.56.29.md:
```diff
@@ -0,0 +1,33 @@
+
+## Improvements
+
+- Absorb `build-nginx` job into `build-images` job (#944) @dhruvkb
```
One downside of having this as a markdown doc and not in an official release is that the reference numbers are no longer auto-linked by GitHub (i.e. `#944` won't go anywhere when this is published to our docs site). Is it possible to update the action to hard-link these?

sarayourfriend replied:
Good point! The easiest way is probably through this sphinx extension: https://github.com/sloria/sphinx-issues

We could also update the change template in release-drafter (https://github.com/release-drafter/release-drafter#change-template-variables) but I think the sphinx plugin would be more broadly useful in the end as it would apply to _all_ docs, not just these ones.

AetherUnbound replied:
Woah, what a cool extension!

sarayourfriend replied:
https://github.com/sloria/sphinx-issues didn't work due to it relying on a custom sphinx role, but I found this SO answer with a handy bit of code that works perfectly for us: https://stackoverflow.com/questions/72424748/myst-parser-auto-linking-linkifying-references-to-bug-tracker-issues

Tested it locally with this PR:

![image](https://user-images.githubusercontent.com/24264157/227411749-2c25a11a-b515-44cc-bd04-8aa52ad1daf4.png)

The "code" was just `Test #987` :tada: 

This is in the latest commit.

AetherUnbound commented on .github/workflows/tag-app-release.yml:
```diff
@@ -0,0 +1,123 @@
+        # and because the action doesn't support generating the notes without
+        # adding them to a release (draft or no), we still let the release get created
+        # and then just copy the body from the result of this action.
+        uses: release-drafter/release-drafter@v5
+        id: release-drafter
+        with:
+          config-name: release_drafter_${{ inputs.app }}.yml
+          version: ${{ steps.tag.outputs.tag }}
```
One concern with this - what does this diff _against_ when it's drafting the release? If it is naively only the last release that was made, we either get in a situation where:
1. Every subsequent changelog includes all previous changes since this workflow was introduced _plus_ the new changes
2. Every changelog only includes the changes _made to the `app` in question since whenever this workflow was last run_, regardless of `app` (e.g. if the runs were `api` -> `frontend` -> `api`, the latter `api` changelog would only have changes from `frontend` -> `api`.

sarayourfriend replied:
Oh jeez, that's a glaring issue with using release-drafter in this way. I believe it would just compare to the previous GitHub release. I think specifically we'd end up with the first case, each changelog including all previous changes for the app in subsequent changelogs.

I think the only way to get around this is going to be to generate the change list ourselves or find another action that allows you to define which commit to start from.

We could also fork release-drafter to add that capability to it ourselves (and while we're add it, the ability to stop it from creating the draft release and instead just build the markdown output).

I'll dig around and see if there's another action we can use or if we'll need to take an independent approach here.

sarayourfriend replied:
Digging more into it, other people use release-drafter with monorepos, so we should be able to do so as well: https://github.com/release-drafter/release-drafter/issues/1294

It looks like `tag-prefix` might tell release drafter how to find the "previous tag" to use when comparing. The docs aren't clear enough on it though, so I'll need to dig and find if that's the case.

sarayourfriend replied:
Yup, after looking at the code it does just that: when it searches for the "latestRelease" it filters the releases by `tagPrefix`: https://github.com/release-drafter/release-drafter/blob/master/lib/releases.js#L57-L59

_However_, this requires you to actually be using GitHub releases.

Which means we have a few options:

1. Go our own way and don't use release-drafter or use a fork of it modified to look at tags generally, not just GitHub releases
  - Only necessary if we're against GitHub release entirely
  - Big downside of needing a lot of extra work
2. Use release drafter the way guardrail is doing so and have it draft releases for each of our three applications. Create a new workflow triggered by releases that parses the release's name to get the name of the app and tags the image and then triggers the production deployment only for that app (by dispatching the corresponding production deployment workflow)
  - Benefit here is that it's pretty simple
  - One downside is that it will still be stuck using the meaningless semantic versioning. We could update the version template to just increment the major version number every time and drop the minor and patch number: https://github.com/release-drafter/release-drafter#projects-that-dont-use-semantic-versioning
  - Another downside is that we wouldn't have the changelogs on the documentation site, only in GitHub releases, in one, singular list (not split by app other than the prefix)
3. A hybrid of the two and the current approach: Stop drafting releases on an ongoing basis and instead use this workflow to create the release and immediately publish it. Automatically dispatch the production deployment with the tag generated. We would rename the workflow to just "Release" and keep the inputs as they are now.
  - Benefits:
    - Changelogs get added to the documentation site and can be split by application (making it easier to see the progression and cadence for individual apps in a list)
    - Lets us use CalVer instead of SemVer which is a more meaningful way of marking our releases
  - The only downside is that it's slightly more work than the second option. However, almost all of that work has already been done in this PR, we just have to make a few tweaks and fixes.

My vote is to do the third one and because it's the one that requires the medium amount of work of the three, I'll go ahead and update the branch do reflect it and then we can decide between it and the second option. To me the first option is a no-go, I just don't see why we would pour that much effort into this for so little benefit compared to the other two.

I'll make the following changes:

1. Remove the "draft release" workflow that currently creates the drafted releases on an ongoing basis for each push to main (this needs to happen anyway for this PR)
2. Change the workflow name to "Deploy production"
3. Change the app input to a choice, as you suggested in another comment
4. Update the release-drafter configuration to tell it to immediately publish the release
5. Dispatch the correct deployment workflow if the app was `api` or `frontend`.

AetherUnbound replied:
Thanks for doing all this investigation! I think your plan of action sounds great, and looks like we'd be prepending the app name anyway with this approach 💃🏼

AetherUnbound reviewed (changes requested):
Generally I love the idea behind this PR, but I have some concerns about the end result of these changes.


Regarding your outstanding questions:
1. I also would prefer to keep git tags. I would prefer to have the tags prepended by the app name rather than appended, that way when viewing the tags in an alphabetically sorted list, it's easy to see when the tags were made for an individual service.
2. I think either having this run as part of the deployment workflows or having this kick off the deployment workflow makes a lot of sense. Would this be called as part of the staging deployment workflow, and then the tag would be used manually for production? That would make the most sense to me, given we usually want to deploy to staging prior to prod.

sarayourfriend commented:
> I also would prefer to keep git tags. I would prefer to have the tags prepended by the app name rather than appended, that way when viewing the tags in an alphabetically sorted list, it's easy to see when the tags were made for an individual service.

Agreed, will do this in the next commits.

> I think either having this run as part of the deployment workflows or having this kick off the deployment workflow makes a lot of sense. Would this be called as part of the staging deployment workflow, and then the tag would be used manually for production? That would make the most sense to me, given we usually want to deploy to staging prior to prod.

It hadn't occurred to me to create the tag during staging: but then what's the point of the changelogs? I think this process is only useful if we're going to have any delay between staging and production release, which right now I think we should without doing live integration testing.

The way I envisioned this being used was:

1. Staging deploys like normal, to a SHA
2. Once we're happy with the set of changes for an app in staging, we copy the SHA for the current deployment and then run this workflow

The only modification I could imagine is to make the production deployment workflows call this new one to generate the tag from the supplied SHA and then also immediately deploy it. That way we wouldn't have to copy the SHA and then also copy the resulting tag out from this new workflow: one less change to accidentally copy the wrong thing somehow.

sarayourfriend commented:
I've implemented the changes from this conversation: https://github.com/WordPress/openverse/pull/987#discussion_r1147015594

Thanks heaps @AetherUnbound for the review. You caught some nasty issues with my first attempt here.

sarayourfriend commented:
@AetherUnbound This workflow would also make the "bump" script in the infrastructure config no longer work for the ingestion server, at least not with its current implementation. We'll need to add some tag-prefix filtering there as well to search for the latest _ingestion server_ specific version from the repo rather than the way it works now.

sarayourfriend commented:
I've removed the changelog files I was using for testing. If you want to see how the whole thing will look on the documentation site, checkout https://github.com/WordPress/openverse/pull/987/commits/905bcbf24fb9bce357ec72c205297d8999509a26 and then run the site locally (`just documentation/live`).

sarayourfriend commented:
Thankfully the testing revealed a number of configuration issues. Everything up until the release-drafter works. However, release-drafter _requires_ the configuration file to live on `main`. As this PR hasn't yet been merged, that's not the case.

Therefore, in the interest of being able to test this, I think we should add the `release-drafter-api.yml` in this PR to the `main` branch in a separate PR. We only need to add one to be able to finish testing.

I'm hoping that the release-drafter part will "just work" but I'm convinced that something about the configuration is ever so slightly off still and will need tweaking.

I'll open the PR to add the api configuration to main for this PR to be able to use it to finish testing.

sarayourfriend commented:
Further testing is blocked until https://github.com/WordPress/openverse/pull/1011 is merged

sarayourfriend commented:
Alright! This is fully working now.

Here's a screenshot of the generated release (I deleted the release as it wasn't a genuine one):

![image](https://user-images.githubusercontent.com/24264157/227431302-795649f7-1281-4d5a-acda-165d83834846.png)

Here's the closed PR that added the changelogs for that fake release: https://github.com/WordPress/openverse/pull/1016

Notice that the changelog only includes API changes, nothing else :tada: 

You can see the release tags added to the `openverse-api` and `openverse-api_nginx` images here: https://github.com/wordpress/openverse/pkgs/container/openverse-api and https://github.com/wordpress/openverse/pkgs/container/openverse-api_nginx

I've deleted the git tags from the repository, but here's the list before I deleted them:

![image](https://user-images.githubusercontent.com/24264157/227431600-8c3752f9-213a-4d35-9338-d5a7d4e64b6f.png)

In short: the new action does what it's supposed to do!

1. Create a git tag for the specified SHA (and only if the SHA corresponds to a published docker image)
2. Add the release tag to the docker image
3. Publish a release (this is an ongoing thing we'll need to allow release drafter to pull only the changes since the last release... it relies explicitly on the GitHub releases, not tags!)
4. Open a PR with the changelogs added to the documentation site

After I remove the testing code, it'll also deploy the production API or frontend with the new docker image tag.

sarayourfriend commented on documentation/changelogs/api/index.md:
````diff
@@ -0,0 +1,10 @@
+
+Changelogs for <https://api.openverse.engineering/>.
+
+```{toctree}
+:glob:
+:titlesonly:
+
+*
+```
````
This applies to all the index files in the changelog directories: the reason for this configuration is that it lists only the titles of the changelogs in ascending order. We might want to reverse this to have it list them in descending order so that the most recent changelog is at the top (as with most changelogs).

AetherUnbound replied:
I definitely find descending-order changelogs to be the most helpful, if that's a quick change to make!

AetherUnbound reviewed (approved):
This is fantastic, thanks for all the effort you put into getting this to work and making sure it functions as expected! I have a bunch of minor notes, but nothing to block this PR (all could be a quick add or fast-follow for this). I'm really excited to see how this new process works 🚀

zackkrida commented on documentation/changelogs/api/index.md:
````diff
@@ -0,0 +1,10 @@
+Changelogs for <https://api.openverse.engineering/>.
+
+```{toctree}
+:glob:
+:titlesonly:
+
````
```suggestion
:glob:
:reverse:
:titlesonly:


```

I _believe_ this is all we need to do to swap the order!

https://github.com/sphinx-doc/sphinx/pull/2527

sarayourfriend replied:
Yup! Should be.

zackkrida commented on documentation/changelogs/frontend/index.md:
````diff
@@ -0,0 +1,10 @@
+Changelogs for <https://openverse.org/>.
+
+```{toctree}
+:glob:
+:titlesonly:
+
````
```suggestion
:glob:
:reverse:
:titlesonly:

```

zackkrida commented on documentation/changelogs/index.md:
````diff
@@ -0,0 +1,18 @@
+workflow.
+
+```{toctree}
+:maxdepth: 1
+:glob:
````
```suggestion
:maxdepth: 1
:glob:
:reverse:

```

zackkrida commented on documentation/changelogs/ingestion_server/index.md:
````diff
@@ -0,0 +1,10 @@
+Changelogs for the internal ingestion server service.
+
+```{toctree}
+:glob:
+:titlesonly:
````
```suggestion
:glob:
:reverse:
:titlesonly:
```

zackkrida reviewed (approved):
This is absolutely amazing. I left a suggestion for reversing the changelogs but I'd love to get this merged and tested on monday.

sarayourfriend commented on documentation/index.md:
```diff
@@ -4,6 +4,8 @@ This documentation is for developers who are building Openverse. If you want to
 use the Openverse API, you should instead refer to the
 [API consumer documentation](https://api.openverse.engineering/).
 
+Test #987
+
```
Yes :sweat_smile:

sarayourfriend commented on .github/workflows/release-app.yml:
```diff
@@ -0,0 +1,165 @@
+        uses: release-drafter/release-drafter@v5
+        id: release-drafter
+        with:
+          config-name: release-drafter-${{ inputs.app }}.yml
```
I don't know if it is possible. The documentation says that it should be in the `.github` folder: https://github.com/release-drafter/release-drafter#action-inputs

I tried to look in the code to see if it would support directories and it's a bit of a rabbit hole that I didn't try to find the end of. If we wanted to keep looking, we'd need to start here and see how this config method resolves the file: https://probot.github.io/api/latest/classes/context.Context.html#config

It also makes it seem like the file should just be in `.github` though.

AetherUnbound replied:
Fair enough 🙂

sarayourfriend commented on documentation/_ext/link_issues.py:
```diff
@@ -0,0 +1,370 @@
+"""Issue linking w/ plain-text autolinking, e.g. #42
```
Is there a specific process for doing so aside from including it in the repository?

AetherUnbound replied:
It seems like with the `myst-parser`, another extension we added, we were able to include it as a package: https://github.com/WordPress/openverse/blob/185d9e3c2bb35b088a083159cd29466a9127b673/documentation/Pipfile#L12

I assumed we might be able to follow a similar path with this, using their installation instructions: https://github.com/ignatenkobrain/sphinxcontrib-issuetracker#installation

sarayourfriend replied:
Gotcha. No we can't use the official package as it doesn't work with modern versions of Sphinx, hence the linked SO answer.
